### PR TITLE
Attach capabilities to DID instances

### PR DESCRIPTION
**Pull request by haardikk21**
# DID CACAO Capabilities - #50

## Description

This PR introduces the functionality to attach a CACAO capability to a DID instance. Either by passing in an optional `capability` parameter in the constructor, or by passing it through an extender function `withCapability(...)` which returns a new DID to use.

We want this to be included so that applications can use CACAO based capabilities authorized by users to attach to their DID and update user's streams on behalf of them.

## How Has This Been Tested?

Added a test in `provider-behavior.test.ts` which tests the `createDagJWS` function with a DID that has an attached capability. Ensures that the `jws` signature has the `cap` header present and that a `capabilityBlock` is also returned from the function.

## Definition of Done

Before submitting this PR, please make sure:

- [x] The work addresses the description and outcomes in the issue
- [x] I have added relevant tests for new or updated functionality
- [x] My code follows conventions, is well commented, and easy to understand
- [x] My code builds and tests pass without any errors or warnings
- [x] I have tagged the relevant reviewers
- [x] I have updated the READMEs of affected packages
- [x] I have made corresponding changes to the documentation
- [x] The changes have been communicated to interested parties

### CODE REVIEW

**Comment by oed** [comment]
The cacao CID needs to go into the protected header of the JWS! Maybe modify the `createJWS` method something like this:
```js
  async createJWS<T = any>(payload: T, options: CreateJWSOptions = {}): Promise<DagJWS> {
    if (this._client == null) throw new Error('No provider available')
    if (this._id == null) throw new Error('DID is not authenticated')
    if (this._capability) {
      options.protected = options.protected || {}
      options.protected.cap = `ipfs://${CID.asCID(this._capability.cid).toString()}`
    }
    const { jws } = await this._client.request('did_createJWS', {
      did: this._id,
      ...options,
      payload,
    })
    return jws
  }
```

**Comment by haardikk21** [comment]
> The cacao CID needs to go into the protected header of the JWS! Maybe modify the `createJWS` method something like this:
> 
> ```ts
>   async createJWS<T = any>(payload: T, options: CreateJWSOptions = {}): Promise<DagJWS> {
>     if (this._client == null) throw new Error('No provider available')
>     if (this._id == null) throw new Error('DID is not authenticated')
>     if (this._capability) {
>       options.protected = options.protected || {}
>       options.protected.cap = `ipfs://${CID.asCID(this._capability.cid).toString()}`
>     }
>     const { jws } = await this._client.request('did_createJWS', {
>       did: this._id,
>       ...options,
>       payload,
>     })
>     return jws
>   }
> ```

Does this also mean I don't need to set it directly on `jws.cap`? Basically replace that behavior with this?

**Comment by oed** [comment]
> Does this also mean I don't need to set it directly on jws.cap? Basically replace that behavior with this?

Yes.

> Also, could you expand a bit on why it is this way? I can do it but don't think I fully understand why

We need to sign over the ocap CID. If we didn't it opens up for manipulation. The `protected` property in a JWS is signed together with the payload.

**Review by oed — COMMENTED** [review]
ugh, looks like my comments did not get though yesterday. These were part of the initial review...

File: src/did.ts [hunk 1]
```diff
@@ -78,12 +79,14 @@ export interface DecryptJWEResult {
 export interface DagJWSResult {
   jws: DagJWS
   linkedBlock: Uint8Array
+  capabilityBlock?: Uint8Array
```
[the remark this answers is not in the document]
Comment (haardikk21) [reply src/did.ts:82]: was trying to be somewhat consistent with the `cap` header - no strong feelings either way

File: src/did.ts [hunk 2]
```diff
@@ -122,6 +139,19 @@ export class DID {
     return this._id
   }
 
+  /**
+   * Attach a capability to the DID instance
+   * @param cap The capability to attach
+   * @returns A new DID instance with the capability attached
+   */
+  withCapability(cap: CacaoBlock): DID {
```
[the remark this answers is not in the document]
Comment (haardikk21) [reply src/did.ts:147]: Agreed, better to just use Cacao, CacaoBlock is deterministic

File: src/did.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (oed) [reply src/did.ts:82]: I would go for `cacaoBlock` for now. We should switch to a CAR file in the future anyhow.

**Review by ukstv — APPROVED** [review]

**Review by oed — APPROVED** [review]
>  I have updated the READMEs of affected packages

Would be nice to have an example in the readme :)

File: README.md [hunk 3]
```diff
@@ -111,6 +111,26 @@ const did = new DID({ resolver: KeyResolver.getResolver() })
...
+const seed = // 32 bytes of entropy, Uint8Array
+const provider = new Ed25519Provider(seed)
+
+const siwe = new SiweMessage("...");
+const cacao = Cacao.fromSiweMessage(siwe);
+
+const did = new DID({ provider, resolver: KeyResolver.getResolver() })
+const didWithCap = did.withCapability(cacao)
+
+const didWithCap2 = new DID({provider, resolver: KeyResolver.getResolver(), capability: cacao})
```
Comment (oed) [inline README.md:131]:
```suggestion
// Add capability to existing DID instance
const did = new DID({ provider, resolver: KeyResolver.getResolver() })
const didWithCap = did.withCapability(cacao)

// Add capability to new DID instance
const didWithCap2 = new DID({provider, resolver: KeyResolver.getResolver(), capability: cacao})
```

File: README.md [hunk 4]
```diff
@@ -111,6 +111,26 @@ const did = new DID({ resolver: KeyResolver.getResolver() })
 await did.resolve('did:key:...')
 ```
 
+### Creating a DID with attached CACAO
+
```
Comment (oed) [inline README.md:115]:
```suggestion
Using CACAO OCAPs to create a Key DID that can sign on behalf of a PKH DID.

```

**Review by oed — APPROVED** [review]
Minor suggestions.

File: README.md [hunk 5]
```diff
@@ -111,7 +111,7 @@ const did = new DID({ resolver: KeyResolver.getResolver() })
 await did.resolve('did:key:...')
 ```
 
-### Creating a DID with attached CACAO
+### Using CACAO OCAPs to create a Key DID that can sign on behalf of a PKH DID.
```
Comment (oed) [inline README.md:114]: Title was good, just meant that we would add more context below!

Comment (haardikk21) [reply README.md:114]: Oh woops okay

Comment (haardikk21) [reply README.md]: done

**Pull request merged**